x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DeprecationRoleDescriptorPreprocessor.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.authz.store;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.metadata.AliasOrIndex;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor.IndicesPrivileges;
+import org.elasticsearch.xpack.core.security.authz.permission.IndicesPermission;
+
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.function.BiConsumer;
+import java.util.function.Predicate;
+
+public final class DeprecationRoleDescriptorPreprocessor implements BiConsumer<Set<RoleDescriptor>, ActionListener<Set<RoleDescriptor>>> {
+
+    // package-private for testing
+    static final String DEPRECATION_STANZA = "Role [{}] grants index privileges over the [{}] alias and not over the [{}] index."
+            + " Granting privileges over an alias and hence granting privileges over all the indices that it points to is deprecated and"
+            + " will be removed in a future version of Elasticsearch. Instead define permissions exclusively on indices or index patterns.";
+
+    private final DeprecationLogger deprecationLogger;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final Set<String> cacheKeys;
+
+    public DeprecationRoleDescriptorPreprocessor(ClusterService clusterService, ThreadPool threadPool,
+                                                 DeprecationLogger deprecationLogger) {
+        this.deprecationLogger = deprecationLogger;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.cacheKeys = Collections.newSetFromMap(Collections.synchronizedMap(new LinkedHashMap<String, Boolean>() {
+            @Override
+            protected boolean removeEldestEntry(final Map.Entry<String, Boolean> eldest) {
+                return size() > 128;
+            }
+        }));
+    }
+
+    @Override
+    public void accept(Set<RoleDescriptor> effectiveRoleDescriptors, ActionListener<Set<RoleDescriptor>> roleBuilder) {
```

Comment:
Do we need to make this Async? It looks like we fork to another thread and then log. So this can just be a function.

Reply:
If it does become synchronous, I think it can just be a `Consumer<Set<RoleDescriptor>>` rather than `BiConsumer<>` (or `Function<>`).
I don't think we want to build a generic method to modify the role descriptors, we just want to be able to do stuff _with_ them.